### ISSUE TITLE
[Question] A method to build a tree of files referenced by a kustomization using SDK

### ISSUE THREAD

**Issue by dan-j** [kubernetes-sigs/kustomize#3678]
TL;DR - Is there any ability, using the SDK, to build a tree of files which are referenced by a root `kustomization.yaml` file, including those from recursively nested `resources:`?

---

I'm looking at building a GitHub Action which will perform a diff of many kustomizations which have changed as part of a PR. There are already some existing actions out there, however they run `kustomize build` against all `kustomization.yaml` files in the repo. What I'm looking to do is only build the kustomizations which have files changed from the PR.

Why am I doing this? We have a monorepo of kustomizations for building a new cluster with non-application software (i.e. cert-manager, prometheus, etc.). Including all our overlays we have a few hundred total possible kustomizations. Building them all for every push to a PR takes quite a while (we use some remote bases so these have to be constantly downloaded).

My idea is to build a tree of every file each `kustomization.yaml` references, and then strip out this tree to only leave the ones which reference files which have been changed in the PR. So for example if I change a base with 3 overlays, my action would detect that it only needs to actually build 4 kustomizations, the three overlays and the base (since the base could be built alone).

What have I got so far? I can use the `types.Kustomization` struct and marshal each `kustomization.yaml`, however I'm still left with the problem of processing each `resources`, `components`, `patches[].path` etc. in order to build the tree. This would be quite time consuming so was wondering if this functionality already exists as part of the public API of the SDK since you must clearly already do something like this in order to build the output?

**Comment by Shell32-Natsu** [comment]
I don't know there is such feature in the kustomize. You may need to invest to develop it.

**Comment by dan-j** [comment]
Would this be a feature the contributors would be willing to accept into kustomize itself? Perhaps under the `kustomize cfg` command group?

If done properly I see it as a very useful CI tool which most users could benefit from.

**Comment by Shell32-Natsu** [comment]
We are in a short of hands but we will consider it.

**Comment by dan-j** [comment]
I can't make any promises but I suspect I'll need to implement something which achieves this in the coming months, so I'll probably approach it as a PR into this project, but also try not to tie it too much into the internal APIs so if it's not accepted it can at least be pulled into it's own tool.

**Comment by Shell32-Natsu** [comment]
@dan-j PRs are welcome. Thanks!
